app/repositories/SessionRepository.scala

```diff
@@ -130,57 +128,109 @@ class DefaultSessionRepository @Inject() (config: AppConfig,
+          Future.successful(Right(
+            ProcessContext(
+              sp.process,
+              sp.answers,
+              sp.labels,
+              sp.flowStack,
+              sp.continuationPool,
+              sp.pageMap,
+              sp.legalPageIds,
+              pageUrl.fold[Option[String]](None){_ =>
+                sp.pageHistory.reverse match {
+                  case x :: y :: xs => Some(y.url)
+                  case _ => None
+                }
+              }
+            )
+          )
+        )
+      }
+    }
+    .recover {
+      case lastError =>
+        logger.error(s"Error $lastError while trying to retrieve process from session repo with _id=$key")
+        Left(DatabaseError)
+    }
+
+  def getUpdateForGET(key: String, pageUrl: Option[String], previousPageByLink: Boolean): Future[RequestOutcome[ProcessContext]] =
+    findAndUpdate(
+      Json.obj("_id" -> key),
+      Json.obj(
+        (List(toFieldPair("$set", Json.obj(toFieldPair(TtlExpiryFieldName, Json.obj("$date" -> Instant.now().toEpochMilli))))) ++
+              pageUrl.fold[List[FieldAttr]](Nil)(url => List("$push" -> Json.obj(PageHistoryKey -> PageHistory(url, Nil))))
+        ).toArray: _*),
+      fetchNewObject = false
+    ).flatMap { r =>
+      r.result[DefaultSessionRepository.SessionProcess]
+      .fold {
+        logger.warn(s"Attempt to retrieve cached process from session repo with _id=$key returned no result, lastError ${r.lastError}")
+        Future.successful(Left(NotFoundError): RequestOutcome[ProcessContext])
+      }{ sp => // SessionProcess returned by findAndUpdate() is intentionally that prior to the update!!
+        pageUrl.fold[Future[RequestOutcome[ProcessContext]]](
+          Future.successful(Right(ProcessContext(sp.process, sp.answers, sp.labels, sp.flowStack, sp.continuationPool, sp.pageMap, Nil, None)))
+        ){url =>
+          sp.pageMap.get(url.drop(sp.process.meta.processCode.length)).fold[Future[RequestOutcome[ProcessContext]]]{
+            logger.warn(s"Attempt to move to unknown page $url")
+            Future.successful(Left(NotFoundError))
+          }{pageDesc =>
+            logger.debug(s"Incoming Page: ${pageDesc.id}, $url, current legalPageIds: ${sp.legalPageIds}")
+            if (sp.legalPageIds.isEmpty || sp.legalPageIds.contains(pageDesc.id)){ // Wild card or fixed list of valid page ids
+              val firstPageUrl: String = s"${sp.process.meta.processCode}${sp.process.startUrl.getOrElse("")}"
+              val (backLink, historyUpdate, flowStackUpdate, labelUpdates) = sessionProcessTransition(url, sp, previousPageByLink, firstPageUrl)
+              val labels: Map[String, Label] = sp.labels ++ labelUpdates.map(l => (l.name -> l)).toMap
+              val legalPageIds = Process.StartStanzaId :: pageDesc.id :: pageDesc.next ++
+                                  backLink.fold[List[String]](Nil)(bl => List(sp.pageMap(bl.drop(sp.process.meta.processCode.length)).id)).distinct
```

Comment:
I suspect "distinct" is only being applied to "(bl => List(sp.pageMap(bl.drop(sp.process.meta.processCode.length)).id))". You need to put some brackets around the whole statement and then apply distinct.

Reply:
done